objects/c932e8/script.lua

```diff
@@ -13,42 +44,161 @@ function onLoad()
+end
+
 function doSpiritSetup(params)
     local color = params.color
     if not Global.getVar("gameStarted") then
         Player[color].broadcast("Please wait for the game to start before pressing button!", "Red")
         return
     end
 
-    local zone = getObjectFromGUID(Global.getVar("elementScanZones")[color])
-    local objs = zone.getObjects()
-    local found = false
-    local serpent = nil
-    for _,obj in pairs(objs) do
-        if obj.guid == "cebe09" then
-            found = true
-            serpent = obj
-            break
-       end
-    end
-    if not found then
+    elementScanZone = Global.getVar("elementScanZones")[color]
+
+    serpent = findSerpent(elementScanZone)
+    if serpent == nil then
         Player[color].broadcast("You have not picked Serpent Slumbering Beneath the Island!", "Red")
         return
     end
 
     local snapPoints = serpent.getSnapPoints()
-
-    table.insert(snapPoints,{position={1.14, 0.20, 0.65}})
-    table.insert(snapPoints,{position={0.92, 0.20, 0.65}})
-    table.insert(snapPoints,{position={0.70, 0.20, 0.65}})
-    table.insert(snapPoints,{position={1.25, 0.20, 0.84}})
-    table.insert(snapPoints,{position={1.03, 0.20, 0.84}})
-    table.insert(snapPoints,{position={0.81, 0.20, 0.84}})
+    for _, position in pairs(slumberPoints) do
+        table.insert(snapPoints,{position=position})
+    end
     serpent.setSnapPoints(snapPoints)
 
-    self.destruct()
+    initialized = true
+    self.clearButtons()
+    self.locked = true
+    self.interactable = false
+    local position = self.getPosition()
+    position.y = -2
+    self.setPosition(position)
+
+    updatePresence()
+    Wait.time(updatePresence, 1, -1)
+end
+
+local function presenceOnIsland(color)
+    local count = 0
+    for _, presence in pairs(getObjectsWithTag("Presence")) do
+        if presence.getName() == color .. "'s Presence" then
+            local hits = upCast(presence, 1, yHeight - presence.getPosition().y)
+            for _, obj in pairs(hits) do
+                if obj.hasTag("Balanced") or obj.hasTag("Thematic") then
+                    local quantity = presence.getQuantity()
+                    if quantity == -1 then
+                        count = count + 1
+                    else
+                        count = count + quantity
+                    end
+                end
+            end
+        end
+    end
+    return count
+end
+
+local function slumberPresence()
+    local count = 0
+    for _, position in pairs(slumberPoints) do
+        local hits = upCastPosSizRot(
+            serpent.positionToWorld(position), -- pos
+            Vector(1,1,1), -- size
+            Vector(0,0,0), -- roation
+            0 -- distance
+        )
+        for _, hit in pairs(hits) do
+            if hit.hasTag("Presence") then
+                count = count + 1
+            end
+        end
+    end
+    return count
+end
+
+function updatePresence()
+    local color = findColor(elementScanZone)
```

Comment:
i might be misremembering things but doesn't the scripting zone have the color saved on the object?

Reply:
The player block does. I've adjusted things to use that.